File: Chapter14/Python/Example2.py
# Chapter 14 - Παραμετρικοί Έλεγχοι για Δύο Πληθυσμούς και ANOVA
# Example 2: Ανάλυση Ζευγαρωτών Δεδομένων (Paired t-test)

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy import stats

np.random.seed(456)
n = 12
before      = np.round(np.random.normal(85, 12, n), 1)
weight_loss = np.round(np.random.normal(3.5, 2, n), 1)
after       = before - weight_loss

weight_data = pd.DataFrame({
    'Subject'   : range(1, n + 1),
    'Before'    : before,
    'After'     : after,
    'Difference': before - after
})
print(weight_data)

# Paired t-test (one-sided: before > after)
t_stat, p_val = stats.ttest_rel(weight_data['Before'], weight_data['After'],
                                alternative='greater')
print(f"\nPaired t-test: t = {t_stat:.4f}, p-value = {p_val:.4f}")

# Manual calculation
d_bar = weight_data['Difference'].mean()
s_d   = weight_data['Difference'].std(ddof=1)
se_d  = s_d / np.sqrt(n)
t_man = d_bar / se_d
p_man = stats.t.sf(t_man, df=n-1)   # one-sided (upper)

print(f"\nManual: d_bar = {d_bar:.3f}, s_d = {s_d:.3f}, SE = {se_d:.3f}")
print(f"        t = {t_man:.4f}, p-value = {p_man:.4f}")

# 95% CI
t_crit = stats.t.ppf(0.975, df=n-1)
ci_lo  = d_bar - t_crit * se_d
ci_hi  = d_bar + t_crit * se_d
print(f"95% CI for mu_d: [{ci_lo:.2f}, {ci_hi:.2f}] kg")

# Normality of differences
w, p_sw = stats.shapiro(weight_data['Difference'])
print(f"\nShapiro-Wilk: W = {w:.4f}, p = {p_sw:.4f}")

# Visualization
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(13, 5))

for i in range(n):
    ax1.plot(['Before', 'After'],
             [weight_data.loc[i, 'Before'], weight_data.loc[i, 'After']],
             'o-', color='steelblue', alpha=0.5)
ax1.set_title(f'Individual Weight Changes\n'
              f'Mean loss = {d_bar:.2f} kg, p = {p_man:.4f}')
ax1.set_ylabel('Weight (kg)')

ax2.hist(weight_data['Difference'], bins=6, color='steelblue',
         edgecolor='white', alpha=0.8)
ax2.axvline(x=0, color='red', linestyle='--', label='H0: mu_d = 0')
ax2.axvline(x=d_bar, color='orange', linestyle='-',
            linewidth=2, label=f'mean = {d_bar:.2f}')
ax2.set_title('Distribution of Differences')
ax2.set_xlabel('Weight Loss (kg)')
ax2.legend()

plt.tight_layout()
plt.savefig('paired_analysis.png', dpi=300, bbox_inches='tight')

File: Chapter14/R/Example2.R
# Chapter 14 - Παραμετρικοί Έλεγχοι για Δύο Πληθυσμούς και ANOVA
# Example 2: Ανάλυση Ζευγαρωτών Δεδομένων (Paired t-test)

set.seed(456)
n <- 12

# Simulate weight data
before       <- round(rnorm(n, mean = 85, sd = 12), 1)
weight_loss  <- round(rnorm(n, mean = 3.5, sd = 2), 1)
after        <- before - weight_loss

weight_data <- data.frame(
  Subject    = 1:n,
  Before     = before,
  After      = after,
  Difference = before - after
)
print(weight_data)

cat("Mean before:", round(mean(weight_data$Before), 2), "kg\n")
cat("Mean after :", round(mean(weight_data$After),  2), "kg\n")
cat("Mean diff  :", round(mean(weight_data$Difference), 2), "kg\n\n")

# Paired t-test (one-sided: loss > 0 means before > after)
paired_test <- t.test(weight_data$Before, weight_data$After,
                      paired = TRUE, alternative = "greater")
print(paired_test)

# Manual calculation
d_bar  <- mean(weight_data$Difference)
s_d    <- sd(weight_data$Difference)
se_d   <- s_d / sqrt(n)
t_stat <- d_bar / se_d
df     <- n - 1
p_val  <- pt(t_stat, df = df, lower.tail = FALSE)

cat("\n--- Manual Calculation ---\n")
cat("d-bar:", round(d_bar, 3), "kg\n")
cat("s_d  :", round(s_d, 3),   "kg\n")
cat("SE_d :", round(se_d, 3),  "kg\n")
cat("t    :", round(t_stat, 3), "\n")
cat("df   :", df, "\n")
cat("p-value (one-sided):", round(p_val, 4), "\n")

# 95% Confidence interval
t_crit <- qt(0.975, df = df)
ci_lo  <- d_bar - t_crit * se_d
ci_hi  <- d_bar + t_crit * se_d
cat("95% CI for mu_d: [", round(ci_lo, 2), ",", round(ci_hi, 2), "] kg\n")

# Normality check on differences
sw <- shapiro.test(weight_data$Difference)
cat("\nShapiro-Wilk on differences: W =", round(sw$statistic, 4),
    ", p =", round(sw$p.value, 4), "\n")

# Visualization: paired plot
library(ggplot2)
library(tidyr)

plot_long <- pivot_longer(weight_data, cols = c(Before, After),
                          names_to = "Time", values_to = "Weight")
plot_long$Time <- factor(plot_long$Time, levels = c("Before", "After"))

ggplot(plot_long, aes(x = Time, y = Weight, group = Subject)) +
  geom_line(color = "steelblue", alpha = 0.5) +
  geom_point(aes(color = Time), size = 3) +
  labs(
    title    = "Individual Weight Changes",
    subtitle = paste("Mean loss:", round(d_bar, 2), "kg,",
                     "p-value =", round(p_val, 4)),
    y = "Weight (kg)", x = ""
  ) +
  theme_minimal(base_size = 13) + theme(legend.position = "right")
